README.md
[![Build Status](https://travis-ci.org/codymorrison/ms-conversion.svg?branch=master)](https://travis-ci.org/codymorrison/ms-conversion) [![Coverage Status](https://coveralls.io/repos/codymorrison/ms-conversion/badge.svg)](https://coveralls.io/r/codymorrison/ms-conversion)


# ms-time

Takes any millisecond output and allows you to output in a number of formats based on the length of time. Supports years, months, weeks, days, minutes, seconds and milliseconds.

## Installation

**NPM:**

```npm i ms-time --save```

## Usage

```js
var msTime = require('ms-time');

msTime(1337000000);
//=> '2w'

msTime(1337000000, { long: true );
//=> '2 weeks'

msTime(1337000000, { weeks: false );
//=> '15.5d'

msTime(1337000000, { long: true, weeks: false );
//=> '15.5 days'
```

## Options

### long

Type: ```Boolean```<br>
Default: ```false```

Outputs in short or long format (i.e. **d** or **days**).

### pluralize

Type: ```Boolean```<br>
Default: ```true```

Setting to false does not pluralize the time, useful for custom formats.

### years

Type: ```Boolean```<br>
Default: ```true```

If the time is greater than 1 year, output in years. If false, fall back to months or next available option.

### months

Type: ```Boolean```<br>
Default: ```true```

If the time is greater than 1 month, output in months. If false, fall back to weeks or next available option.

### weeks

Type: ```Boolean```<br>
Default: ```true```

If the time is greater than 1 week, output in weeks. If false, fall back to days or next available option.

### days

Type: ```Boolean```<br>
Default: ```true```

If the time is greater than 1 day, output in days. If false, fall back to hours.

### seconds

Type: ```Boolean```<br>
Default: ```true```

If set to false and time is less than 1 minute, this will output a time of 1 minute.

### milliseconds

Type: ```Boolean```<br>
Default: ```true```

If set to false and time is less than 1 second, this will output a time of 1 second.

### yearFormat

Type: ```Array```<br>
Default: ```['y', 'year']```

Customize the output of years, first value is the short value, the second is the long. Neither should be plural.

### monthFormat

Type: ```Array```<br>
Default: ```['mo', 'month']```

Customize the output of months, first value is the short value, the second is the long. Neither should be plural.

### weekFormat

Type: ```Array```<br>
Default: ```['w', 'week']```

Customize the output of weeks, first value is the short value, the second is the long. Neither should be plural.

### dayFormat

Type: ```Array```<br>
Default: ```['d', 'day']```

Customize the output of days, first value is the short value, the second is the long. Neither should be plural.

### hourFormat

Type: ```Array```<br>
Default: ```['h', 'hour']```

Customize the output of hours, first value is the short value, the second is the long. Neither should be plural.

### minuteFormat

Type: ```Array```<br>
Default: ```['m', 'minute']```

Customize the output of minutes, first value is the short value, the second is the long. Neither should be plural.

### secondFormat

Type: ```Array```<br>
Default: ```['s', 'second']```

Customize the output of seconds, first value is the short value, the second is the long. Neither should be plural.

### millisecondFormat

Type: ```Array```<br>
Default: ```['ms', 'millisecond']```

Customize the output of milliseconds, first value is the short value, the second is the long. Neither should be plural.

## Contributing

1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Commit your changes: `git commit -am 'Add some feature'`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request :D


## Credits

Originally based on the exellent library [ms.js](https://github.com/rauchg/ms.js). Also relies on [ms-parse](https://github.com/sindresorhus/pretty-ms).

## License

MIT


History.md
0.4.0 / 2015-04-28
==================

  * Add ```yearFormat``` option
  * Add ```monthFormat``` option
  * Add ```weekFormat``` option
  * Add ```dayFormat``` option
  * Add ```hourFormat``` option
  * Add ```secondFormat``` option
  * Add ```millisecondFormat``` option

0.3.1 / 2015-04-28
==================

  * Add ```pluralization``` option

0.3.0 / 2015-04-27
==================

  * Add ```seconds``` option
  * Add ```minuteFormat``` option

0.2.0 / 2015-04-27
==================

  * Add ```milliseconds``` option

0.1.0 / 2015-04-26
==================

  * Initial release


test/test.js

/**
 * Dependencies.
 */

if ('undefined' != typeof require) {
  expect = require('expect.js');
  ms = require('../');
}

// numbers

describe('ms(number, { long: true })', function() {
  it('should support milliseconds', function() {
    expect(ms(500, { long: true })).to.be('500 milliseconds');
  });

  it('should support seconds', function() {
    expect(ms(1000, { long: true })).to.be('1 second');
    expect(ms(1200, { long: true })).to.be('1 second');
    expect(ms(10000, { long: true })).to.be('10 seconds');
  });

  it('should support minutes', function() {
    expect(ms(60 * 1000, { long: true })).to.be('1 minute');
    expect(ms(60 * 1200, { long: true })).to.be('1 minute');
    expect(ms(60 * 10000, { long: true })).to.be('10 minutes');
  });

  it('should support hours', function() {
    expect(ms(60 * 60 * 1000, { long: true })).to.be('1 hour');
    expect(ms(60 * 60 * 1200, { long: true })).to.be('1 hour');
    expect(ms(60 * 60 * 10000, { long: true })).to.be('10 hours');
  });

  it('should support days', function() {
    expect(ms(24 * 60 * 60 * 1000, { long: true })).to.be('1 day');
    expect(ms(24 * 60 * 60 * 1200, { long: true })).to.be('1 day');
    expect(ms(24 * 60 * 60 * 10000, { long: true, weeks: false })).to.be('10 days');
  });

  it('should support weeks', function() {
    expect(ms((24 * 60 * 60 * 1000) * 7, { long: true })).to.be('1 week');
    expect(ms((24 * 60 * 60 * 1000) * 11, { long: true })).to.be('1.5 weeks');
    expect(ms((24 * 60 * 60 * 1000) * 35, { long: true, months: false})).to.be('5 weeks');
  });

  it('should support months', function() {
    expect(ms(24 * 60 * 60 * 1000 * 30, { long: true })).to.be('1 month');
    expect(ms(24 * 60 * 60 * 1000 * 45 , { long: true })).to.be('1.5 months');
    expect(ms(24 * 60 * 60 * 1000 * 30 * 14, { long: true, years: false })).to.be('14 months');
  });

  it('should support years', function() {
    expect(ms(24 * 60 * 60 * 1000 * 365, { long: true })).to.be('1 year');
    expect(ms(24 * 60 * 60 * 1000 * 520, { long: true })).to.be('1.5 years');
    expect(ms(24 * 60 * 60 * 1000 * 900, { long: true })).to.be('2.5 years');
  });

  it('should allow you to change the format of all types', function() {
    expect(ms(1000 * 60 * 60 * 24 * 365, { yearFormat: ['min', 'max'] })).to.be('1min');
    expect(ms(1000 * 60 * 60 * 24 * 30, { monthFormat: ['min', 'max'] })).to.be('1min');
    expect(ms(1000 * 60 * 60 * 24 * 7, { weekFormat: ['min', 'max'] })).to.be('1min');
    expect(ms(1000 * 60 * 60 * 24, { dayFormat: ['min', 'max'] })).to.be('1min');
    expect(ms(1000 * 60 * 60, { hourFormat: ['min', 'max'] })).to.be('1min');
    expect(ms(1000 * 60, { minuteFormat: ['min', 'max'] })).to.be('1min');
    expect(ms(1000, { secondFormat: ['min', 'max'] })).to.be('1min');
    expect(ms(1, { millisecondFormat: ['min', 'max'] })).to.be('1min');
  });

  it('should default to 1 minute if < 1 minute and seconds are disabled', function() {
    expect(ms(30000, { seconds: false})).to.be('1m');
  });

  it('should default to 1 second if < 1 second and milliseconds are disabled', function() {
    expect(ms(500, { milliseconds: false})).to.be('1s');
  });

  it('should be able to set pluralization', function() {
    expect(ms(60 * 1000 * 47.5, { long: true, pluralize: false })).to.be('47.5 minute');
    expect(ms(60 * 1000 * 47.5, { long: true, pluralize: true })).to.be('47.5 minutes');
  });

  it('should round', function() {
    expect(ms(234234234, { long: true })).to.be('2.5 days');
  });
});

// numbers

describe('ms(number)', function() {
  it('should support milliseconds', function() {
    expect(ms(500)).to.be('500ms');
  });

  it('should support seconds', function() {
    expect(ms(1000)).to.be('1s');
    expect(ms(10000)).to.be('10s');
  });

  it('should support minutes', function() {
    expect(ms(60 * 1000)).to.be('1m');
    expect(ms(60 * 10000)).to.be('10m');
  });

  it('should support hours', function() {
    expect(ms(60 * 60 * 1000)).to.be('1h');
    expect(ms(60 * 60 * 10000)).to.be('10h');
  });

  it('should support days', function() {
    expect(ms(24 * 60 * 60 * 1000, { long: true })).to.be('1 day');
    expect(ms(24 * 60 * 60 * 1200, { long: true })).to.be('1 day');
    expect(ms(24 * 60 * 60 * 10000, { long: true, weeks: false })).to.be('10 days');
  });

  it('should support weeks', function() {
    expect(ms((24 * 60 * 60 * 1000) * 7, { long: true })).to.be('1 week');
    expect(ms((24 * 60 * 60 * 1000) * 11, { long: true })).to.be('1.5 weeks');
    expect(ms((24 * 60 * 60 * 1000) * 35, { long: true, months: false})).to.be('5 weeks');
  });

  it('should support months', function() {
    expect(ms(24 * 60 * 60 * 1000 * 30, { long: true })).to.be('1 month');
    expect(ms(24 * 60 * 60 * 1000 * 45 , { long: true })).to.be('1.5 months');
    expect(ms(24 * 60 * 60 * 1000 * 30 * 14, { long: true, years: false })).to.be('14 months');
  });

  it('should support years', function() {
    expect(ms(24 * 60 * 60 * 1000 * 365, { long: true })).to.be('1 year');
    expect(ms(24 * 60 * 60 * 1000 * 520, { long: true })).to.be('1.5 years');
    expect(ms(24 * 60 * 60 * 1000 * 900, { long: true })).to.be('2.5 years');
  });

  it('should allow you to change the format of all types', function() {
    expect(ms(1000 * 60 * 60 * 24 * 365, { long: true, yearFormat: ['min', 'max'] })).to.be('1 max');
    expect(ms(1000 * 60 * 60 * 24 * 30, { long: true, monthFormat: ['min', 'max'] })).to.be('1 max');
    expect(ms(1000 * 60 * 60 * 24 * 7, { long: true, weekFormat: ['min', 'max'] })).to.be('1 max');
    expect(ms(1000 * 60 * 60 * 24, { long: true, dayFormat: ['min', 'max'] })).to.be('1 max');
    expect(ms(1000 * 60 * 60, { long: true, hourFormat: ['min', 'max'] })).to.be('1 max');
    expect(ms(1000 * 60, { long: true, minuteFormat: ['min', 'max'] })).to.be('1 max');
    expect(ms(1000, { long: true, secondFormat: ['min', 'max'] })).to.be('1 max');
    expect(ms(1, { long: true, millisecondFormat: ['min', 'max'] })).to.be('1 max');
  });

  it('should default to 1 minute if < 1 minute and seconds are disabled', function() {
    expect(ms(30000, { long: true, seconds: false })).to.be('1 minute');
  });

  it('should default to 1 second if < 1 second and milliseconds are disabled', function() {
    expect(ms(500, { long: true, milliseconds: false })).to.be('1 second');
  });

  it('should round', function() {
    expect(ms(234234234)).to.be('2.5d');
  });
});


index.js
var parseMs = require('parse-ms');

/**
 * Config
 */
var config = {
  long: false,
  pluralize: true,
  days: true,
  weeks: true,
  months: true,
  years: true,
  seconds: true,
  milliseconds: true,
  yearFormat: ['yr', 'year'],
  monthFormat: ['mo', 'month'],
  weekFormat: ['w', 'week'],
  hourFormat: ['h', 'hour'],
  dayFormat: ['d', 'day'],
  minuteFormat: ['m', 'minute'],
  secondFormat: ['s', 'second'],
  millisecondFormat: ['ms', 'millisecond']
};

/**
 * Parse or format the given `val`.
 *
 * Options:
 *
 *  - `long` verbose formatting [false]
 *
 * @param {String|Number} val
 * @param {Object} options
 * @return {String|Number}
 * @api public
 */

module.exports = function(val, opts) {
  opts = opts || {};

  config.long = opts.long ? true : false;
  config.pluralize = opts.pluralize === false ? false : true;
  config.years = opts.years === false ? false : true;
  config.months = opts.months === false ? false : true;
  config.weeks = opts.weeks === false ? false : true;
  config.days = opts.days === false ? false : true;
  config.seconds = opts.seconds === false ? false : true;
  config.milliseconds = opts.milliseconds === false ? false : true;
  config.yearFormat = opts.yearFormat ? opts.yearFormat : ['yr', 'year'];
  config.monthFormat = opts.monthFormat ? opts.monthFormat : ['mo', 'month'];
  config.weekFormat = opts.weekFormat ? opts.weekFormat : ['w', 'week'];
  config.dayFormat = opts.dayFormat ? opts.dayFormat : ['d', 'day'];
  config.hourFormat = opts.hourFormat ? opts.hourFormat : ['h', 'hour'];
  config.minuteFormat = opts.minuteFormat ? opts.minuteFormat : ['m', 'minute'];
  config.secondFormat = opts.secondFormat ? opts.secondFormat : ['s', 'second'];
  config.millisecondFormat = opts.millisecondFormat ? opts.millisecondFormat : ['ms', 'millisecond'];

  return formatMs(val);
};

/**
 * Format helper.
 *
 * @param {Number} ms The time to convert in ms
 * @return {String}
 * @api private
 */
function formatMs(ms) {
  var msParsed = parseMs(ms);

  msParsed = addParseMsData(msParsed);

  // year format
  if (msParsed.years > 0 && config.years) return addFormat(msParsed.years, config.yearFormat);

  // month format
  if (msParsed.months > 0 && config.months) return addFormat(msParsed.months, config.monthFormat);

  // week format
  if (msParsed.weeks > 0 && config.weeks) return addFormat(msParsed.weeks, config.weekFormat);

  // day format
  if (msParsed.days > 0 && config.days) return addFormat(msParsed.days, config.dayFormat);

  // hour format
  if (msParsed.hours > 0) return addFormat(msParsed.hours, config.hourFormat);

  // minute format
  if (msParsed.minutes > 0) return addFormat(msParsed.minutes, config.minuteFormat);

  // second format
  if (msParsed.seconds > 0) {
    if (!config.seconds) return (addFormat(1, config.minuteFormat));

    return addFormat(msParsed.seconds, config.secondFormat);
  }

  // ms format
  if (msParsed.milliseconds > 0) {
    if (!config.milliseconds) return addFormat(1, config.secondFormat);

    return addFormat(msParsed.milliseconds, config.millisecondFormat);
  }

  return 0;
}


/**
 * Add Format
 *
 * @param {Float} time The finalized time
 * @param {Array} format The format to use for short/long
 * @return {String}
 * @api private
 */
function addFormat(time, format) {
  var usePlural = '';
  var formatType = config.long ? 1 : 0;
  var space = config.long ? ' ' : '';

  if (time > 1 && config.long && config.pluralize) usePlural = 's';

  return time + space + format[formatType] + usePlural;
}


/**
 * Round helper
 *
 * Takes a value and divides it by a number returning the
 * value in .5 increments.
 *
 * @param {Number} dividend The time to divide from
 * @param {Number} divisor The number to divide by
 * @param {Number} +/-
 * @return {Float} The rounded dividend (0, 0.5, 1) from current number
 * @api private
 */
function roundHelper(dividend, divisor, scale) {
  scale = scale || 0.1;

  // The Number of times the dividend goes into the divisor
  var integer = Math.floor(dividend / divisor);

  // This gets the decimal point and collects the first 2 numbers (0.0 - 1.0)
  var fraction = (dividend / divisor) % 1;

  // If fraction is greater than scale return ceiling of integer
  if (fraction >= (0.5 + scale)) {
    return integer++;
  }

  // If fraction is less than scale return floor of integer
  if (fraction <= (0.5 - scale)) {
    return integer;
  }

  // Otherwise we're within scale and can apply 0.5
  return integer + 0.5;
}


/**
 * Round Days
 *
 * Round days using hours and day data
 *
 * @param {Number} days The number of days
 * @param {Number} hours The number of hours
 * @param {Number} scale
 * @return {Float} The day amount in 0.5 increments based on scale
 * @api private
 */
function roundDays(days, hours, scale) {
  scale = scale || 6;
  var median = 12;

  if (hours >= (median + scale)) {
    return days++;
  }

  if (hours <= (median - scale)) {
    return days;
  }

  return days + 0.5;
}


/**
 * Round Hours
 *
 * Round hours using minutes and hour data
 *
 * @param {Number} hours The time of hours
 * @param {Number} minutes The number of minutes
 * @param {Number} scale
 * @return {Float} The hour amount in 0.5 increments based on scale
 * @api private
 */
function roundHours(hours, minutes, scale) {
  scale = scale || 15;
  var median = 30;

  if (minutes >= (median + scale)) {
    return hours++;
  }

  if (minutes <= (median - scale)) {
    return hours;
  }

  return hours + 0.5;
}


/**
 * Round Minutes
 *
 * Round minutes using minute and seconds data
 *
 * @param {Number} minutes The time of seconds
 * @param {Number} seconds The number of minutes
 * @param {Number} scale
 * @return {Float} The minute amount in 0.5 increments based on scale
 * @api private
 */
function roundMinutes(minutes, seconds, scale) {
  scale = scale || 15;
  var median = 30;

  if (seconds >= (median + scale)) {
    return minutes++;
  }

  if (seconds <= (median - scale)) {
    return minutes;
  }

  return minutes + 0.5;
}


/**
 * addParseMsData
 *
 * Adds additional values when days are greater than 7
 *
 * @param {Object} msData The parse-ms data
 * @return {Object} The data including years/months/days
 * @api private
 */
function addParseMsData(msData) {
  var actualHours = (msData.days * 24) + msData.hours;
  var actualMinutes = (actualHours * 60) + msData.minutes;

  msData.years = 0;
  msData.months = 0;
  msData.weeks = 0;

  // Calculate Years
  if (msData.days >= 365) msData.years = roundHelper(msData.days, 365, null);

  // Calculate Months
  if (msData.days >= 30) msData.months = roundHelper(msData.days, 30, null);

  // Calculate Weeks
  if (msData.days >= 7) msData.weeks = roundHelper(msData.days, 7, null);

  if (msData.days > 0) msData.days = roundDays(msData.days, msData.hours, 6);
  if (msData.hours > 0) msData.hours = roundHours(actualHours, msData.minutes, 15);
  if (msData.minutes > 0) msData.minutes = roundMinutes(actualMinutes, msData.seconds, 15);

  return msData;
}
